### ISSUE TITLE
Fill entire screen when fullscreen on MacBook with Notch?

### ISSUE THREAD

**Issue by shelbyKiraM** [iina/iina#4083]
**What you want IINA to do:**
Have option to allow the full screen view to fill the entire display area like e.g. video in Telegram overlay:
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/202692/202053511-ceac1bb4-f74b-496a-a60f-9813048219a9.png">
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/202692/202053555-6ca73e9a-8759-4d60-9e71-c089fc62707a.png">

**What IINA does currently:**
doesn't expand into notch area in full screen (notice black line at top; I use autohide menu bar, but it works in TG either way):
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/202692/202053733-9e46e1c4-e2ce-48a9-ac00-1dc3680efbf8.png">

**Why you think this should be added:**
so I can enable for 3:2 or 4:3 vid to let it be larger at a small loss of the notch covering part for improved watchability.

**Examples of other projects that have something similar:**
[Telegram Desktop](https://github.com/telegramdesktop/tdesktop) macOS was where I found it working right.

**Comment by shelbyKiraM** [comment]
Found the below via [https://appleinsider.com/articles/21/10/19/apps-can-use-full-screen-area-around-new-macbook-pros-camera-notch](https://appleinsider.com/articles/21/10/19/apps-can-use-full-screen-area-around-new-macbook-pros-camera-notch):

> According to newly published HIG [documents](https://developer.apple.com/documentation/bundleresources/information_property_list/nsprefersdisplaysafeareacompatibilitymode), MacBooks that include a camera housing in the screen bezel provide a special operating mode that prevents full screen apps from placing content under the notch. When active, compatibility mode automatically changes the active area of the system's display to avoid the housing cutout, ensuring that content is not obscured.
> 
> Developers [can opt](https://developer.apple.com/design/human-interface-guidelines/macos/app-architecture/fullscreen-mode/) for a custom full-screen experience, but they will need to define safe areas in their code using new tools. For example, app makers can specify a large safe area with an inset at the top of the screen that leaves the entirety of the app unobscured. Alternatively, new properties can be set to identify rectangular portions of screen to the right and left of the notch, areas that will be earmarked for active content.

**Comment by low-batt** [comment]
The "special operating mode" referred to in the Apple Insider article is controlled by the [NSPrefersDisplaySafeAreaCompatibilityMode](https://developer.apple.com/documentation/bundleresources/information_property_list/nsprefersdisplaysafeareacompatibilitymode) key:
> On Mac computers that include a camera housing, the Finder adds a checkbox to your app’s Get Info panel to enable or disable compatibility mode. If your app’s Info.plist file includes the `NSPrefersDisplaySafeAreaCompatibilityMode` key, the Finder doesn’t add that checkbox; instead, it uses the value of the key to determine whether to run your app in compatibility mode. Set the value of the key to true to always run your app in compatibility mode, and set it to false to never run your app in compatibility mode.

In the latest IINA releases `NSPrefersDisplaySafeAreaCompatibilityMode` is set to false. The safe area compatibility mode is disabled. Finder will not display the checkbox in the Get Info panel for IINA.

The applicable section of Apple's Human Interface Guidelines is the [Going full screen](https://developer.apple.com/design/human-interface-guidelines/patterns/going-full-screen) referred to in the Apple Insider article. Under macOS that document provides the follow guidance:
> **If you use custom full-screen support, rely on system-defined areas to keep your full-screen window content unobscured.** For example, some Mac models include a camera housing that occupies an area at the top-center of the screen. Using the system’s full-screen support automatically accommodates this area; if you use a custom full-screen experience, you need to account for system-defined areas as you position your content.

IINA does provide custom full screen support. It is controlled by the `Use legacy full screen` preference found in the `General` tab of preferences in the `Behavior` section. IINA is using the [safeAreaInsets](https://developer.apple.com/documentation/appkit/nsscreen/3882821-safeareainsets) referred to by the HIG to avoid having content obscured by the camera housing. So a lot of the code required to support this proposed feature is already present.

However, despite the HIG indicating that applications can implement their own custom full screen support, changes in macOS Big Sur badly broke IINA's custom support in multiple ways resulting in crashes if the  `Use legacy full screen` preference was enabled. After a bunch of workarounds for AppKit issues were added to the code this feature is somewhat working. But there are still issues with it that need to be solved.

So the priority for developers this area of IINA is to resolve the remaining problems and get legacy full screen fully working like it was before macOS Big Sur.

**Comment by beyondcompute** [comment]
I would very much like to upvote having an option to fill the entire screen in full-screen mode on a “notched” monitor. 🖥️👍

**Comment by eiko4** [comment]
@beyondcompute +1

**Comment by Brog33** [comment]
Wonder if this would be easier to implement for sonoma

**Comment by Rozzemak** [comment]
After building current dev iina in fleet and removing the safe range offset in the _setWindowFrameForLegacyFullScreen_ method, it works just as expected in for example 4:3 ratio movies on notched screen. It's beautiful. Really should be an option.

<img width="1588" alt="image" src="https://github.com/iina/iina/assets/28568671/3975e0c3-272a-4e19-9631-95a19dc3cb73">

**Comment by low-batt** [comment]
Correct. It is more work to add the UI for something like a `Confine to safe area` subordinate checkbox to the existing `Use legacy full screen` setting than it is to implement the behavior.

**Comment by kishan-dhankecha** [comment]
Is this still under active consideration for implementing?

**Comment by low-batt** [comment]
On this:
> Is this still under active consideration for implementing?

The issue is still open, so yes.

There had been a disincentive to address issues requiring changes to the settings window because major changes were underway in how that area of the IINA UI was coded along with changes to how that window looks and operates. While this was going on changes to settings required double the work, as both the old code and the new code had to be modified. The new code has been merged into the `develop` branch although the new settings window has not yet been enabled as there are a few problems left to address. But the incentive to wait on changes involving settings is no longer there.

I will ask the senior developers about this issue. A bit hard to get peoples attention right now. Currently developers are very busy. Some UI changes that have been wanted for a long time, but were postponed due to the major changes required, are now being implemented.  The focus is currently on finishing those changes and finding and fixing regressions introduced by the changes. I just happened to notice yesterday that there is currently a regression in legacy full screen in the development version of IINA that needs to be looked into.

Work is also underway to upgrade the libraries from other projects that IINA is dependent upon, especially mpv and FFmpeg.

One thing I find disturbing is that the text I quoted [above](https://github.com/iina/iina/issues/4083#issuecomment-1339964458) from the [Going full screen](https://developer.apple.com/design/human-interface-guidelines/going-full-screen) section of Apple's Human Interface Guidelines has been changed and no longer refers to "custom full-screen support". IINA's custom full screen support still has workarounds for the AppKit problems Apple introduced in macOS Big Sur that broke IINA's legacy full screen implementation. Hopefully Apple won't cause more problems for applications with their own full screen implementation.
